spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
         valid = True
         return valid, message
 
+    def apply_plugin_settings(self, options):
+        """
+        Determine what to do to apply configuration plugin settings.
+        """
+        pass
 
-# =============================================================================
-# SpyderPluginWidget
-# =============================================================================
-class BasePluginWidget(QWidget, BasePluginWidgetMixin):
-    """
-    Basic functionality for Spyder plugin widgets.
+    def update_font(self):
+        """
+        This must be reimplemented by plugins that need to adjust
+        their fonts.
+        """
+        pass
 
-    WARNING: Don't override any methods or attributes present here!
-    """
 
-    # Signal used to update the plugin title when it's undocked
-    sig_update_plugin_title = Signal()
+class SpyderPlugin(QObject, PluginBaseMixin):
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
 
-    def __init__(self, main=None):
-        super(BasePluginWidget, self).__init__(main)
+    # Use this signal to display a message in the status bar.
+    # str: The message you want to display
+    # int: Amount of time to display the message
+    sig_show_status_message = Signal(str, int)
 
-        # Dockwidget for the plugin, i.e. the pane that's going to be
-        # displayed in Spyder for this plugin.
-        # Note: This is created when you call the `add_dockwidget`
-        # method, which must be done in the `register_plugin` one.
-        self.dockwidget = None
+    # Use this signal to inform another plugin that configuration changed.
+    sig_option_changed = Signal(str, object)
 
-    def add_dockwidget(self):
-        """Add the plugin's QDockWidget to the main window."""
-        super(BasePluginWidget, self)._add_dockwidget()
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent)
+        self._initialize(parent, configuration)
 
-    def tabify(self, core_plugin):
+    # FIXME: consider renaming this to setup_plugin, so we can achieve some
+    # language consistency. With this we can then do post_setup_plugin
+    # to refer to things to do after setup, or when visible.
+    # Using register and then post_visible_setup is not clear.
```

Comment:
I'm -0.5 on this. It'll change a lot of stuff just for some more consistency. At moment  there are no plugin doing something after this step (if I recall things correctly). That's because post_visible_setup is only done in the main window.

Reply:
>I'm -0.5 on this. It'll change a lot of stuff just for some more consistency. 

Changes won't break anything, they will add the proper deprecation methods for removal on the next major version. I do not plan to keep breaking plugins among versions but we do need to have consistency as that makes developing plugins easier (which is the purpose of this).

>That's because post_visible_setup is only done in the main window.

Some plugins may want to do something after the setup is ready and want to defer it to the post_visible setup method. Like for instance Kite is doing on first install. Right now a signal is used, but we have only that signal which is odd. So it is much better to provide a logic for plugins: 

For example

```python
def register():
    # Create actions and connect to signals, register any plugin types.
    # For example adding a new type to the variable explorer

def setup():
    # Prepare the plugin to start during load

def after_setup():
    # Prepare the plugin to do something after spyder is visble and fully loaded

def unregister():
    # If we wanted to add support for disabling a plugin without a spyder restart this method
    # would perform cleanup and disconnection
```

Now on the main window we can do something like

```
python

def register():
    for plugin in plugins:
        try:
             plugin.register()

def setup():
    for plugin in plugins:
        try:
             plugin.setup()

def after_setup():
    for plugin in plugins:
        try:
             plugin.after_setup()

```

In this process all the hard coded code in main window like defining hard coded plugins and actions ane menus will be moved to the plugin that actually does it or needs it.

This will reduce mainwindow complexity and generalize the process that is currently doing a lot of "manual" work.

Reply:
@jitseniesen well we don't need. Is just to keep things more consistent.

We could only have setup and post_setup or something... 

But register probably sounds better. I just want to think of things that make the api consistent

Reply:
Sorry for being dense, but what is currently inconsistent?